ls1intum/Artemis
# `Exam mode`: Update cache logic of the exam live statistics

### Checklist
#### General


#### Server

### Motivation and Context



During local extensive testing with more than 10000 actions (per single student) I noticed that some actions are not saved in the cache. The reason is that the write operation for each student was not properly locked. So if a student performs several actions in less than a second, there is a possibility that the actions are not saved.

### Description


To avoid locking the entire exam cache, I added an additional method that locks only the exam activities and not the entire exam. Therefore, we can insert multiple actions from different users without locking.


### Steps for Testing

Prerequisites:
- 1 Instructor
- Multiple Students
- 1 Exam with Exam Live Statistics enabled

1. Log in to Artemis
2. Create an Exam with Exam Live Statistics enabled
3. Register multiple Students
4. Participate with those students
5. View the Exam Live Statistics as instructor during the Exam
6. Reload the page
8. Verify that the provided information are correct and the old actions are also loaded


### Review Progress





#### Code Review
#### Manual Tests
